File: ipad-app/DrawingToolbar.swift
// 펜, 형광펜, 지우개, 올가미와 각 도구의 세부 설정을 보여 주는 툴바입니다.
// 실제 그리기 처리는 PencilCanvas가 맡고 이 화면은 DrawTool 값만 바꿉니다.

import SwiftUI

// 항상 보이는 필기 툴바 — 펜/형광펜/지우개/올가미 + 색·두께.
struct DrawingToolbar: View {
    @Binding var tool: DrawTool

    private let widths: [CGFloat] = [2, 5, 9]
    private let eraserWidths: [CGFloat] = [12, 24, 40]

    var body: some View {
        ScrollView(.horizontal, showsIndicators: false) {
        HStack(spacing: 10) {
            HStack(spacing: 4) {
                toolButton(.pen, "pencil.tip", "펜")
                toolButton(.marker, "highlighter", "형광펜")
                toolButton(.eraser, "eraser", "지우개")
                toolButton(.lasso, "lasso", "올가미")
            }

            if tool.kind.usesColor {
                Divider().frame(height: 26)

                HStack(spacing: 6) {
                    ForEach(DrawTool.palette, id: \.self) { hex in
                        Circle()
                            .fill(Color(uiColor: UIColor(hex: hex)))
                            .frame(width: 22, height: 22)
                            .overlay(
                                Circle().stroke(Color.primary, lineWidth: tool.colorHex == hex ? 2 : 0)
                            )
                            .onTapGesture { tool.colorHex = hex }
                    }
                    ColorPicker(
                        "사용자 색상",
                        selection: Binding(
                            get: { Color(uiColor: UIColor(hex: tool.colorHex)) },
                            set: { tool.colorHex = UIColor($0).hexRGB }
                        ),
                        supportsOpacity: false
                    )
                    .labelsHidden()
                    .frame(width: 28)
                }

                Divider().frame(height: 26)

                HStack(spacing: 8) {
                    ForEach(widths, id: \.self) { w in
                        widthDot(dot: w + 4, selected: tool.width == w) { tool.width = w }
                    }
                }
            } else if tool.kind == .eraser {
                Divider().frame(height: 26)

                HStack(spacing: 4) {
                    eraserModeButton(.pixel, "일부")
                    eraserModeButton(.object, "전체")
                }

                Divider().frame(height: 26)

                HStack(spacing: 8) {
                    ForEach(eraserWidths, id: \.self) { w in
                        // 지우개 실제 굵기는 최대 40이라 점 크기는 보기 좋게 눌러 표시.
                        widthDot(dot: w / 2 + 4, selected: tool.eraserWidth == w) { tool.eraserWidth = w }
                    }
                }
            }
        }
        }
        .padding(8)
        .background(.regularMaterial, in: RoundedRectangle(cornerRadius: 14))
        .padding(.horizontal, 16)
    }

    private func eraserModeButton(_ mode: EraserMode, _ label: String) -> some View {
        Button {
            tool.eraserMode = mode
        } label: {
            Text(label)
                .font(.system(size: 13, weight: .medium))
                .frame(width: 40, height: 30)
                .background(
                    tool.eraserMode == mode ? Color.accentColor.opacity(0.2) : .clear,
                    in: RoundedRectangle(cornerRadius: 8)
                )
        }
        .buttonStyle(.plain)
    }

    private func widthDot(dot: CGFloat, selected: Bool, action: @escaping () -> Void) -> some View {
        Circle()
            .fill(Color.primary)
            .frame(width: dot, height: dot)
            .frame(width: 30, height: 30)
            .background(selected ? Color.primary.opacity(0.12) : .clear, in: Circle())
            .onTapGesture(perform: action)
    }

    private func toolButton(_ kind: PenKind, _ icon: String, _ label: String) -> some View {
        Button {
            tool.kind = kind
        } label: {
            Image(systemName: icon)
                .font(.system(size: 17))
                .frame(width: 40, height: 34)
                .background(
                    tool.kind == kind ? Color.accentColor.opacity(0.2) : .clear,
                    in: RoundedRectangle(cornerRadius: 8)
                )
        }
        .buttonStyle(.plain)
        .accessibilityLabel(label)
    }
}

File: flutter-app/lib/widgets/drawing_toolbar.dart
import 'package:flutter/material.dart';

import '../canvas/drawing_controller.dart';
import '../canvas/stroke.dart';

/// 항상 보이는 필기 툴바 — 펜/형광펜/지우개/올가미 + 색·두께.
class DrawingToolbar extends StatelessWidget {
  const DrawingToolbar({super.key, required this.controller});

  final DrawingController controller;

  static const _widths = [2.0, 5.0, 9.0];
  static const _eraserWidths = [12.0, 24.0, 40.0];

  DrawTool get _tool => controller.tool;

  @override
  Widget build(BuildContext context) {
    return AnimatedBuilder(
      animation: controller,
      builder: (context, _) => Material(
        color: Theme.of(context).colorScheme.surface.withValues(alpha: 0.92),
        borderRadius: BorderRadius.circular(14),
        elevation: 3,
        child: Padding(
          padding: const EdgeInsets.all(8),
          child: Row(
            mainAxisSize: MainAxisSize.min,
            children: [
              _toolButton(PenKind.pen, Icons.edit),
              _toolButton(PenKind.marker, Icons.brush),
              _toolButton(PenKind.eraser, Icons.cleaning_services_outlined),
              _toolButton(PenKind.lasso, Icons.highlight_alt),
              if (_tool.usesColor) ...[
                _divider(),
                for (final c in DrawTool.palette) _colorDot(c),
                _divider(),
                for (final w in _widths)
                  _widthDot(
                    dot: w + 4,
                    selected: _tool.width == w,
                    onTap: () {
                      _tool.width = w;
                      controller.toolChanged();
                    },
                  ),
              ] else if (_tool.kind == PenKind.eraser) ...[
                _divider(),
                _eraserModeButton(EraserMode.pixel, '일부'),
                _eraserModeButton(EraserMode.object, '전체'),
                _divider(),
                for (final w in _eraserWidths)
                  _widthDot(
                    // 지우개 실굵기는 최대 40이라 점 크기는 눌러 표시.
                    dot: w / 2 + 4,
                    selected: _tool.eraserWidth == w,
                    onTap: () {
                      _tool.eraserWidth = w;
                      controller.toolChanged();
                    },
                  ),
              ] else if (_tool.kind == PenKind.lasso &&
                  controller.selection.isNotEmpty) ...[
                // 선택된 획 색 변경 + 삭제.
                _divider(),
                for (final c in DrawTool.palette) _colorDot(c),
                _divider(),
                InkWell(
                  borderRadius: BorderRadius.circular(8),
                  onTap: controller.deleteSelected,
                  child: const SizedBox(
                    width: 40,
                    height: 34,
                    child: Icon(Icons.delete_outline, size: 19),
                  ),
                ),
              ],
            ],
          ),
        ),
      ),
    );
  }

  Widget _divider() => Container(
        width: 1,
        height: 26,
        margin: const EdgeInsets.symmetric(horizontal: 8),
        color: const Color(0x33888888),
      );

  Widget _toolButton(PenKind kind, IconData icon) {
    final selected = _tool.kind == kind;
    return InkWell(
      borderRadius: BorderRadius.circular(8),
      onTap: () {
        _tool.kind = kind;
        controller.clearSelection();
        controller.toolChanged();
      },
      child: Container(
        width: 40,
        height: 34,
        decoration: BoxDecoration(
          color: selected ? const Color(0x332F6BFF) : null,
          borderRadius: BorderRadius.circular(8),
        ),
        child: Icon(icon, size: 19),
      ),
    );
  }

  Widget _colorDot(Color color) {
    final selected = _tool.color == color;
    return GestureDetector(
      onTap: () {
        // 올가미 선택 중이면 선택 획의 색을 바꾼다.
        if (_tool.kind == PenKind.lasso) {
          controller.recolorSelected(color);
          return;
        }
        _tool.color = color;
        controller.toolChanged();
      },
      child: Container(
        width: 22,
        height: 22,
        margin: const EdgeInsets.symmetric(horizontal: 3),
        decoration: BoxDecoration(
          color: color,
          shape: BoxShape.circle,
          border: selected ? Border.all(width: 2, color: Colors.black87) : null,
        ),
      ),
    );
  }

  Widget _widthDot({
    required double dot,
    required bool selected,
    required VoidCallback onTap,
  }) {
    return GestureDetector(
      onTap: onTap,
      child: Container(
        width: 30,
        height: 30,
        alignment: Alignment.center,
        decoration: BoxDecoration(
          color: selected ? const Color(0x1F000000) : null,
          shape: BoxShape.circle,
        ),
        child: Container(
          width: dot,
          height: dot,
          decoration: const BoxDecoration(
            color: Colors.black87,
            shape: BoxShape.circle,
          ),
        ),
      ),
    );
  }

  Widget _eraserModeButton(EraserMode mode, String label) {
    final selected = _tool.eraserMode == mode;
    return InkWell(
      borderRadius: BorderRadius.circular(8),
      onTap: () {
        _tool.eraserMode = mode;
        controller.toolChanged();
      },
      child: Container(
        width: 40,
        height: 30,
        alignment: Alignment.center,
        decoration: BoxDecoration(
          color: selected ? const Color(0x332F6BFF) : null,
          borderRadius: BorderRadius.circular(8),
        ),
        child: Text(label,
            style: const TextStyle(fontSize: 13, fontWeight: FontWeight.w500)),
      ),
    );
  }
}
